### ISSUE TITLE
Not working with Jruby in SSL mode

### ISSUE THREAD

**Issue by atharva-infiverve** [WinRb/WinRM#249]
Hi

We are using winrm 2.1.1 with Jruby 9.1.1.0. 

Here is the script we are trying to run:
```ruby
require 'winrm'
opts = {
  endpoint: 'https://some.hostname.com:5986/wsman',
  transport: :ssl,
  no_ssl_peer_verification: :true,
  user: 'administrator',
  password: 'some-password'
}
conn = WinRM::Connection.new(opts)
conn.shell(:powershell) do |shell|
  output = shell.run('$PSVersionTable') do |stdout, stderr|
    STDOUT.print stdout
    STDERR.print stderr
  end
  puts "The script exited with exit code #{output.exitcode}"
end
```
We are constantly getting the following error. Please note that this works well in Negotiate and Plaintext mode.

Thanks a lot for your help. Following is the stacktrace we are getting: 

NoMethodError: undefined method `to_der' for #<HTTPClient::JRubySSLSocket::JavaCertificate:0x56aaaecd>
                 channel_hash at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:52
             application_data at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:59
  gss_channel_bindings_struct at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:45
        channel_binding_token at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:36
                  target_info at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:227
                         blob at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:217
                 nt_proof_str at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:208
  calculate_user_session_key! at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:196
                authenticate! at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:27
                 init_context at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client.rb:36
                    init_auth at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/http/transport.rb:229
                 send_request at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/http/transport.rb:168
         max_envelope_size_kb at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:152
       max_fragment_blob_size at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:80
                   fragmenter at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:198
  block in open_shell_payload at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:162
                          map at org/jruby/RubyArray.java:2300
           open_shell_payload at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:159
                   open_shell at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:119
                block in open at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:163
                    retryable at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/retryable.rb:37
                         open at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:161
           with_command_shell at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:124
                          run at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:80
      block in winrm_test2.rb at winrm_test2.rb:11
                        shell at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/connection.rb:43
                        <top> at winrm_test2.rb:10
atharva@venus:~/Documents/flint-banglore/flint-1.2.0.0/flintbox/example$ ruby --2.2 winrm_test2.rb
warning: --2.2 ignored
NoMethodError: undefined method `to_der' for #<HTTPClient::JRubySSLSocket::JavaCertificate:0x56aaaecd>
                 channel_hash at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:52
             application_data at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:59
  gss_channel_bindings_struct at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:45
        channel_binding_token at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:36
                  target_info at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:227
                         blob at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:217
                 nt_proof_str at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:208
  calculate_user_session_key! at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:196
                authenticate! at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:27
                 init_context at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/rubyntlm-0.6.1/lib/net/ntlm/client.rb:36
                    init_auth at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/http/transport.rb:229
                 send_request at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/http/transport.rb:168
         max_envelope_size_kb at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:152
       max_fragment_blob_size at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:80
                   fragmenter at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:198
  block in open_shell_payload at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:162
                          map at org/jruby/RubyArray.java:2300
           open_shell_payload at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:159
                   open_shell at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/power_shell.rb:119
                block in open at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:163
                    retryable at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/retryable.rb:37
                         open at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:161
           with_command_shell at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:124
                          run at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/shells/base.rb:80
      block in winrm_test2.rb at winrm_test2.rb:11
                        shell at /home/atharva/.rvm/gems/jruby-9.0.5.0/gems/winrm-2.1.2/lib/winrm/connection.rb:43
                        <top> at winrm_test2.rb:10

**Comment by sneal** [comment]
It looks like there's some compatibility issues between rubyntlm gem and the JRuby runtime. We don't explicitly test or support JRuby, but might be willing to accept a patch that addresses your issue.

**Comment by atharva-infiverve** [comment]
Hi @sneal,
Thanks. We are investigating the issue but we have not worked with rubyntlm before .So it would be great if you could suggest some steps to fix the issue so that we can take this forward and also do the testing with Jruby.

**Comment by mwrock** [comment]
@atharva-infiverve I just took a peak at the `JavaCertificate` implementation in `HTTPClient`. I don't have a jruby environment easily accessible but it would be interesting if you manually edited `lib/winrm/http/transport.rb` line 228 to:
```
binding = r.peer_cert.nil? ? nil : Net::NTLM::ChannelBinding.create(r.peer_cert.cert)
```
Note this just adds `.cert` to `r.peer_cert` which I believe gets the underlying X509 cert that should have the `to_der` method.

If tha tdoes work, a PR could be to check if `peer_cert` is a `JavaCertificate` and if so, sent its `.cert`.

**Comment by atharva-infiverve** [comment]
@mwrock,

Hi,Sorry for the delay. We tried with the solution suggested by you and edited the file : `lib/winrm/http/transport.rb` line 228 to: 

`binding = r.peer_cert.nil? ? nil : Net::NTLM::ChannelBinding.create(r.peer_cert.cert)`

But we got the following stacktrace:

```
NoMethodError: undefined method `to_der' for #<Java::SunSecurityX509::X509CertImpl:0x789aa6ab>
                 channel_hash at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:52
             application_data at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:59
  gss_channel_bindings_struct at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:45
        channel_binding_token at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/channel_binding.rb:36
                  target_info at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:227
                         blob at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:217
                 nt_proof_str at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:208
  calculate_user_session_key! at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:196
                authenticate! at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/client/session.rb:27
                 init_context at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/rubyntlm-0.6.1/lib/net/ntlm/client.rb:36
                    init_auth at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/http/transport.rb:229
                 send_request at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/http/transport.rb:168
         max_envelope_size_kb at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/power_shell.rb:152
       max_fragment_blob_size at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/power_shell.rb:80
                   fragmenter at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/power_shell.rb:198
  block in open_shell_payload at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/power_shell.rb:162
                          map at org/jruby/RubyArray.java:2355
           open_shell_payload at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/power_shell.rb:159
                   open_shell at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/power_shell.rb:119
                block in open at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/base.rb:163
                    retryable at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/retryable.rb:37
                         open at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/base.rb:161
           with_command_shell at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/base.rb:124
                          run at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/shells/base.rb:80
              block in (root) at example:winrm_test2.rb:11
                        shell at /home/atharva/Documents/flint-banglore/flint-1.2.0.0/lib/gems/gems/winrm-2.1.1/lib/winrm/connection.rb:43
```
So when we replaced the .to_der method  to .getEncoded() in file:
 lib/net/ntlm/channel_binding.rb line no 52 to:

@channel_hash ||= OpenSSL::Digest::SHA256.new(channel.getEncoded()) 
the issue got fixed for  Jruby 9.1.1.0.                     

So we need to fix two patches in order to execute the gems in Jruby:

1. Edit the file : `lib/winrm/http/transport.rb` line 228 to: 

if defined? JRUBY_VERSION:

`binding = r.peer_cert.nil? ? nil : Net::NTLM::ChannelBinding.create(r.peer_cert.cert)`

else keep it unchanged:

`binding = r.peer_cert.nil? ? nil : Net::NTLM::ChannelBinding.create(r.peer_cert)

2. Edit the file :  lib/net/ntlm/channel_binding.rb line no 52 to:
if defined? JRUBY_VERSION:

@channel_hash ||= OpenSSL::Digest::SHA256.new(channel.getEncoded()) 

else keep it unchanged :

@channel_hash ||= OpenSSL::Digest::SHA256.new(channel.to_der)

Thanks alot for your solution. It worked with few path fixes mentioned above.

**Comment by mwrock** [comment]
I wonder if changing
```
binding = r.peer_cert.nil? ? nil : Net::NTLM::ChannelBinding.create(r.peer_cert.cert)
```
to
```
cert = OpenSSL::X509::Certificate.new(r.peer_cert.cert.getEncoded())
binding = r.peer_cert.nil? ? nil : Net::NTLM::ChannelBinding.create(cert)
```
works?

That way we can isolate the jruby handling to this gem.

**Comment by mwrock** [comment]
Hmm. I don't see the PR. I only see 1 open PR in this repo.

**Comment by atharva-infiverve** [comment]
Hello @mwrock ,
Can you please check and confirm it now.

**Comment by pcai** [comment]
closing as resolved by #272
